### PR TITLE
[checkpointio] sharded optimizers checkpointing feature adapted for DDP plugin

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#3986

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Adapted the newly developed feature: sharded optimizers checkpointing to TorchDDPPlugin.
Tidied up arguments passing in several booster IO methods.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/booster.py [hunk 1]
```diff
@@ -179,24 +180,34 @@ def load_model(self, model: nn.Module, checkpoint: str, strict: bool = True):
...
 
     def save_model(self,
-                   model: nn.Module,
+                   model: Union[nn.Module, ModelWrapper],
                    checkpoint: str,
-                   prefix: str = None,
                    shard: bool = False,
-                   size_per_shard: int = 1024):
+                   gather_dtensor: bool = True,
+                   prefix: str = None,
```
Comment (ver217) [inline colossalai/booster/booster.py:187]: type should be `Optional[str]`

File: colossalai/booster/booster.py [hunk 2]
```diff
@@ -205,22 +216,34 @@ def load_optimizer(self, optimizer: Optimizer, checkpoint: str):
...
         self.checkpoint_io.load_optimizer(optimizer, checkpoint)
 
-    def save_optimizer(self, optimizer: Optimizer, checkpoint: str, shard: bool = False, size_per_shard: int = 1024):
-        """Save optimizer to checkpoint.
-        Warning: Saving sharded optimizer checkpoint is not supported yet.
+    def save_optimizer(self,
+                       optimizer: Optimizer,
+                       checkpoint: str,
+                       shard: bool = False,
+                       gather_dtensor=True,
```
Comment (ver217) [inline colossalai/booster/booster.py:229]: Please add type hint

File: colossalai/booster/plugin/torch_ddp_plugin.py [hunk 3]
```diff
@@ -62,8 +62,12 @@ def save_sharded_model(self,
         if self.coordinator.is_master():
             super().save_sharded_model(model, checkpoint_path, gather_dtensor, prefix, max_shard_size, use_safetensors)
 
-    def save_sharded_optimier(self, optimizer: Optimizer, checkpoint: str, gather_dtensor: bool, prefix: str,
-                              size_per_shard: int):
+    def save_sharded_optimier(self,
+                              optimizer: Optimizer,
+                              checkpoint: str,
+                              gather_dtensor: bool = False,
```
Comment (ver217) [inline colossalai/booster/plugin/torch_ddp_plugin.py:68]: I find that the default value of `gather_dtensor` of difference methods are different. Please use the same value.

File: colossalai/booster/plugin/torch_ddp_plugin.py [hunk 4]
```diff
@@ -62,8 +62,12 @@ def save_sharded_model(self,
...
+                              prefix: str = None,
```
Comment (ver217) [inline colossalai/booster/plugin/torch_ddp_plugin.py:69]: Wrong type hint

File: colossalai/checkpoint_io/utils.py [hunk 5]
```diff
@@ -139,6 +140,8 @@ def shard_optimizer_checkpoint(state_dict: dict, max_shard_size: int = 1024) ->
         state_size = 0
         isDTensor = False
         for state_tensor in state.values():
+            if state_tensor is None:
```
Comment (ver217) [inline colossalai/checkpoint_io/utils.py:143]: In which case will the `state_tensor` be None?

File: colossalai/checkpoint_io/utils.py [hunk 6]
```diff
@@ -311,10 +318,24 @@ def cast(param, value, key=None):
         else:
             new_states[k] = v
 
-    optimzier.state.update(new_states)
+    if isinstance(optimizer, OptimizerWrapper):
+        # If optimizer is wrapped, unwrap it.
+        optimizer = optimizer.optim
```
Comment (ver217) [inline colossalai/checkpoint_io/utils.py:323]: It would be better to unwrap it outside

File: colossalai/checkpoint_io/utils.py (hunk 5 above)
Comment (Fridge003) [reply colossalai/checkpoint_io/utils.py:143]: The state 'momentum_buffer' of SGD optimizer can be None. In my case if I delete this two lines, the code will fail the test for ddp_checkpoint & fsdp_checkpoint.

**Review by ver217 — APPROVED** [review]

**Pull request merged**